supervisor/iptablesctrl/uidpam/update_portsets.go

```diff
@@ -0,0 +1,260 @@
+package uipam
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"os/user"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/aporeto-inc/trireme-lib/cache"
+	"github.com/bvandewalle/go-ipset/ipset"
+	"go.uber.org/zap"
+)
+
+const (
+	procNetTCPFile                    = "/proc/net/tcp"
+	portSetUpdateIntervalMilliseconds = 1000
+	portEntryTimeout                  = 60
+	uidFieldOffset                    = 7
+	procHeaderLineNum                 = 1
+	portOffset                        = 1
+	ipPortOffset                      = 1
+	sockStateOffset                   = 3
+	sockListeningState                = "0A"
+	hexFormat                         = 16
+	integerSize                       = 64
+	decFormat                         = 10
+)
+
+// PortSet : This provides an interface to update the
+// look up table required to program the ipset portsets.
+type PortSet interface {
+	AddToUIDToPortSetCache(uid string, value string) (err error)
+	GetFromUIDToPortSetCache(uid string) (err error)
+	AddToUIDToPortsCache(uid string, value string) (ok bool)
+	RemoveFromUIDToPortSetCache(uid string) (err error)
+	RemoveFromUIDToPortsCache(uid string) (err error)
+	UpdatePortSet(uid string, port string)
+}
+
+// UIDCacheInstance : This type contains look up tables
+// to help update the ipset portsets.
+type UIDCacheInstance struct {
+	UIDToPortSet *cache.Cache
+	UIDToPorts   *cache.Cache
+	UIDSet       map[string]bool
+}
+
+// NewInstance : creates a new UIDCache instance
+func NewInstance() (p *UIDCacheInstance, err error) {
+
+	p = &UIDCacheInstance{
+		UIDToPortSet: cache.NewCache("UIDToPortSet"),
+		UIDToPorts:   cache.NewCache("UIDToPorts"),
+		UIDSet:       make(map[string]bool), // updated on portset create/delete
+	}
+
+	return p, nil
+}
+
+func (p *UIDCacheInstance) getUserID(username string) (string, error) {
+	u, err := user.Lookup(username)
+	if err != nil {
+		return "", err
+	}
+	return u.Uid, nil
+}
+
+// AddToUIDToPortsCache This adds/updates UIDToPorts cache. This cache
+// maps UID to list of active listening ports.
+func (p *UIDCacheInstance) AddToUIDToPortsCache(username string, portStr string) (ok bool, err error) {
+	ok = false
+	portList := make([]int64, 0)
+	portNum, err := strconv.ParseInt(portStr, decFormat, integerSize)
+
+	if err != nil {
+		return ok, err
+	}
+
+	uid, err := p.getUserID(username)
+	if err != nil {
+		return ok, err
+	}
+
+	if v, err := p.UIDToPortSet.Get(uid); err == nil {
+		portList = append(v.([]int64), portNum)
+		p.UIDToPorts.AddOrUpdate(uid, portList)
+		ok = true
+	} else {
+		portList = append(portList, portNum)
+		p.UIDToPorts.Add(uid, portList)
+	}
+
+	return ok, nil
+}
+
+// AddToUIDToPortSetCache : This adds/updates UIDToPortSet cache. This cache
```

Comment:
What kind of name is this - Add To UID To PortSet Cache ? Why not AddToUIDPortSetCache ? Why is this cache needed ?

Reply:
Names Changed.